### PR TITLE
Add guide for encrypting data at datastore layer

### PR DESCRIPTION
**Pull request by php-coder**
This is a guide for enabling data encryption at datastore layer. It is based on the similar [Kubernetes guide](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/) from @smarterclayton but it has been adapted for OpenShift.

PTAL @smarterclayton @openshift/team-documentation @openshift/sig-security

Trello: https://trello.com/c/yhuQZ3NP/84-5-security-sccfsi-encrypt-secrets-at-rest-in-etcd

### CODE REVIEW

File: _topic_map.yml [hunk 1]
```diff
@@ -616,6 +616,9 @@ Topics:
   - Name: Sysctls
     File: sysctls
     Distros: openshift-origin,openshift-enterprise
+  - Name: Encrypting Data at Rest
+    File: encrypting_data
+    Distros: openshift-origin,openshift-enterprise,openshift-dedicated
```
Comment (php-coder) [inline _topic_map.yml]: Is this list correct? Most of the entries in this section don't include openshift-dedicated and I don't know why.

File: admin_guide/encrypting_data.adoc [hunk 2]
```diff
@@ -0,0 +1,264 @@
+[[admin-guide-encrypting-data-rest]]
+= Encrypting Data at Rest
+{product-author}
+{product-version}
+:data-uri:
+:icons:
+:experimental:
```
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: I don't know whether I need `:experimental:` and `{product-author}`, `{product-version}`. I've copied them over from another topic but I don't know what they are doing.

File: admin_guide/encrypting_data.adoc [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+oc create secret generic secret1 -n default --from-literal=mykey=mydata
+----
+
+. Using the etcdctl commandline, read that secret out of etcd:
++
+----
+ETCDCTL_API=3 etcdctl get /kubernetes.io/secrets/default/secret1 [...] | hexdump -C
+----
++
+where *[...]* must be the additional arguments for connecting to the etcd server.
```
Comment (soltysh) [inline admin_guide/encrypting_data.adoc]: That's very vague. It would be nice to point to an appropriate document showing what the other flags are. The only one I found is this: https://docs.openshift.org/latest/admin_guide/backup_restore.html

File: _topic_map.yml [hunk 4]
```diff
@@ -616,6 +616,9 @@ Topics:
   - Name: Sysctls
     File: sysctls
     Distros: openshift-origin,openshift-enterprise
+  - Name: Encrypting Data at Rest
```
Comment (soltysh) [inline _topic_map.yml]: REST is the appropriate spelling, it's short for Representational state transfer

File: admin_guide/encrypting_data.adoc [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+.Excerpt of *_master-config.yaml_*
+====
+
+[source,yaml]
+----
+kubernetesMasterConfig:
+  apiServerArguments:
+    experimental-encryption-provider-config:
+    - /path/to/encryption.config
```
Comment (soltysh) [inline admin_guide/encrypting_data.adoc]: Again, explain the main lines here.

File: admin_guide/encrypting_data.adoc [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+. Remove the old decryption key from the config after you back up etcd with the new key in use and
+update all secrets
+
+With a single API server step 2 may be skipped
+
+
+[[encrypting-data-decryption]]
+== Decrypting Data
+
+To disable encryption at rest place the *identity* provider as the first entry in the config:
```
Comment (soltysh) [inline admin_guide/encrypting_data.adoc]: REST

File: admin_guide/encrypting_data.adoc [hunk 7]
```diff
@@ -0,0 +1,261 @@
+[[admin-guide-encrypting-data-rest]]
+= Encrypting Data at Rest
```
Comment (soltysh) [inline admin_guide/encrypting_data.adoc]: REST

File: admin_guide/encrypting_data.adoc [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+toc::[]
+
+[[encrypting-data-overview]]
+== Overview
+
+This topic shows how to enable and configure encryption of secret data at rest.
+
+[[encrypting-data-configuration]]
+== Configuration and Determining Whether Encryption at Rest is Already Enabled
```
Comment (soltysh) [inline admin_guide/encrypting_data.adoc]: REST

File: admin_guide/encrypting_data.adoc [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+----
+kubernetesMasterConfig:
+  apiServerArguments:
+    experimental-encryption-provider-config:
+    - /path/to/encryption.config
+----
+====
+
+[[encrypting-data-encryption-config]]
+== Understanding the Encryption at Rest Configuration
```
Comment (soltysh) [inline admin_guide/encrypting_data.adoc]: REST and all other places as well, please.

File: _topic_map.yml (hunk 4 above)
Comment (php-coder) [inline _topic_map.yml]: This is not about REST but about "at rest". I'll try to find a better wording to make it clear.

File: admin_guide/encrypting_data.adoc (hunk 2 above)
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: I looked at the 7 files I have open and none uses those fields. Not sure if we use them.

File: admin_guide/encrypting_data.adoc (hunk 8 above)
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: Do we need to define what _at rest_ means?

File: admin_guide/encrypting_data.adoc [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+[source,yaml]
+----
+kind: EncryptionConfig
+apiVersion: v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - identity: {}
+    - aesgcm:
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: We use <#> annotations in code sample. 
Put <1> <2> <3>... at the end of a line and below the sample <1> description.
- aesgcm:<1>

<1> This parameter does this.

File: admin_guide/encrypting_data.adoc [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        keys:
+        - name: key1
+          secret: YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=
+----
+====
+
+Each *resources* array item is a separate config and contains a complete configuration. The
+*resources.resources* field is an array of Kubernetes resource names (*resource* or *resource.group*)
+that should be encrypted. The *providers* array is an ordered list of the possible encryption
+providers. Only one provider type may be specified per entry (*identity* or *aescbc* may be provided,
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: s/*resources*/`resources`
s/*providers*/`providers` 

Or, I can do a follow-up PR.

File: admin_guide/encrypting_data.adoc (hunk 10 above)
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: The table below probably works better than <1>. Maybe mention "See the table below for descriptions of the parameters in the following sample code."

File: admin_guide/encrypting_data.adoc [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+| *aesgcm*
+| AES-GCM with random nonce
+| Must be rotated every 200k writes
+| Fastest
+| 16, 24, or 32-byte
+| Is not recommended for use except when an automated key rotation scheme is implemented.
+
+|===
+
+Each provider supports multiple keys - the keys are tried in order for decryption, and if the
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: s/keys - the/keys. The

Again, I can do a follow on PR for formatting things.

File: admin_guide/encrypting_data.adoc [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+        keys:
+        - name: key1
+          secret: <BASE 64 ENCODED SECRET>
+    - identity: {}
+----
+
+To create a new secret perform the following steps:
+
+. Generate a 32 byte random key and base64 encode it. If you are on Linux or Mac OS X, run the
+following command:
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: Can you use another OS other than Linux or Mac OS? 
If not, you can take them out. Might make people think they can.
If so, do we have the command for other OS?

**Comment by mburke5678** [comment]
@php-coder A few comments. I can fix most of the comments I made in a follow-up PR if you want.

File: admin_guide/encrypting_data.adoc [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
++
+where *[...]* must be the additional arguments for connecting to the etcd server.
+
+. Verify the stored secret is prefixed with *k8s:enc:aescbc:v1:* which indicates the *aescbc*
+provider has encrypted the resulting data.
+
+. Verify the secret is correctly decrypted when retrieved via the API:
++
+----
+oc describe secret generic -n default
```
Comment (pweil-) [inline admin_guide/encrypting_data.adoc]: s/generic/secret1

File: admin_guide/encrypting_data.adoc [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+----
++
+should match *mykey: mydata*
```
Comment (pweil-) [inline admin_guide/encrypting_data.adoc]: this doesn't actually show the data, I see:

```
$ oc describe secret secret1 -n default
Name:		secret1
Namespace:	default
Labels:		<none>
Annotations:	<none>

Type:	Opaque

Data
====
mykey:	6 bytes
```

could use something like 
```
$ oc get secret secret1 -n default --template='{{index .data "mykey"}}' | base64 -d
mydata
```

File: admin_guide/encrypting_data.adoc (hunk 3 above)
Comment (pweil-) [inline admin_guide/encrypting_data.adoc]: here is an example if it is helpful

```
 ETCDCTL_API=3 etcdctl --endpoints="https://192.168.1.118:4001" --cacert="./secret-test/master/ca-bundle.crt" --cert=./secret-test/master/admin.crt --key=./secret-test/master/admin.key get /kubernetes.io/secrets/default/secret1 | hexdump -C
2017-08-20 18:51:28.146273 I | warning: ignoring ServerName for user-provided CA for backwards compatibility is deprecated
00000000  2f 6b 75 62 65 72 6e 65  74 65 73 2e 69 6f 2f 73  |/kubernetes.io/s|
00000010  65 63 72 65 74 73 2f 64  65 66 61 75 6c 74 2f 73  |ecrets/default/s|
00000020  65 63 72 65 74 31 0a 6b  38 73 3a 65 6e 63 3a 61  |ecret1.k8s:enc:a|
00000030  65 73 63 62 63 3a 76 31  3a 6b 65 79 31 3a 06 f5  |escbc:v1:key1:..|
00000040  95 61 b6 5a fb db ca 16  d8 89 6c 70 60 81 7d 0d  |.a.Z......lp`.}.|
00000050  4a 8b b9 63 25 be e7 f9  0a e0 2e 98 ba cb 39 f2  |J..c%.........9.|
00000060  30 28 a0 c1 6e 97 a8 63  ba 11 ae 1c 8d 69 35 9c  |0(..n..c.....i5.|
00000070  c4 18 5c e5 df 03 a5 43  1f 70 26 54 ad 34 ed e9  |..\....C.p&T.4..|
00000080  69 63 3f 05 96 28 a0 53  1d 4e 2e f1 46 1d 22 99  |ic?..(.S.N..F.".|
00000090  2b 15 34 87 bf 1c 93 42  1e 9c a8 26 2a b5 fd ce  |+.4....B...&*...|
000000a0  28 f8 42 1b ea e7 0b ad  c7 26 49 c5 4a cd 69 18  |(.B......&I.J.i.|
000000b0  fc 9d ae 40 79 60 4e 34  cd 08 50 d7 46 96 88 5e  |...@y`N4..P.F..^|
000000c0  d8 8e c7 d6 a7 ff 2a f9  8a 50 41 7b 23 92 0a     |......*..PA{#..|
000000cf
```

File: admin_guide/encrypting_data.adoc (hunk 14 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: Good catch! Will be fixed.

Also I've submitted it to the upstream: https://github.com/kubernetes/kubernetes.github.io/pull/5134

File: admin_guide/encrypting_data.adoc (hunk 13 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: Users can generate random bytes on any OS  but the command will be different of course. This command was tested on Mac OS X/Linux but I wouldn't surprise if it also works on FreeBSD, for example (actually it should work).

So, how to express this? "For example, on Linux and Mac OS X the following command can be used:" is it better?

Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: How about: "Users can generate random bytes on any OS. For example, on Linux and Mac OS X the following command can be used:"

File: admin_guide/encrypting_data.adoc (hunk 15 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: I'd like to use this command but it won't work on Mac OS X where we need to use `base64 -D` for decoding.

What's about this?

> `oc get secret secret1 -n default -o yaml | grep mykey`
> should match `mykey: bXlkYXRh`

File: admin_guide/encrypting_data.adoc [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+| *secretbox*
+| XSalsa20 and Poly1305
+| Strong
+| Faster
+| 32-byte
+| A newer standard and may not be considered acceptable in environments that require high levels of
+review.
+
+| *aesgcm*
+| AES-GCM with random nonce
```
Comment (tiran) [inline admin_guide/encrypting_data.adoc]: ``random nonce`` is an oxymoron. A value can't be completely random and a nonce at the same time. Random values can be repeated, a nonce must never be repeat itself.

Kubernetes has no proper nonce generator and uses a random IV as nonce for AES-GCM. Since AES-GCM requires a proper nonce to be secure, AES-GCM is not recommended. The 200k write limit just limits the possibility of a fatal nonce misuse to a reasonable low margin.

File: admin_guide/encrypting_data.adoc (hunk 8 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: @tiran This violates our docs guide: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#topic-titles-and-section-headings

File: admin_guide/encrypting_data.adoc (hunk 16 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: @tiran Thanks! So, what I should write here then?

File: admin_guide/encrypting_data.adoc (hunk 8 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: > Do we need to define what at rest means?

@mburke5678 I'm going to completely replace it by "at the data store layer" to reduce possible misunderstanding.

File: admin_guide/encrypting_data.adoc (hunk 16 above)
Comment (abstractj) [inline admin_guide/encrypting_data.adoc]: @tiran only my 2 cents. I believe the reason why it was written as "random nonce", it's because that's not obvious for newcomers and it aligns with the documentation here https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/

Comment (simo5) [inline admin_guide/encrypting_data.adoc]: php-coder: "AES-GCM with a random IV" add a note about a random IV being used instead of a proper nonce and how this maybe a problem therefore not recommended ?

**Comment by php-coder** [comment]
Thank you all for the feedback! I've updated document to include all these comments.

Also all occurrences of "at rest" were replaced by "at datastore layer" as per this dialog with @smarterclayton:
```
<semushin> claytonc, what is the differences here between "encrypted at the datastore layer" and "encrypted at REST"?
<semushin> claytonc, looks like you've implemented the first but in many places we use the second
<claytonc> HA, that should be "at rest"
<claytonc> not "at REST"
<claytonc> we encrypt over REST already
<semushin> claytonc, the secrets encrypted at REST by using HTTPS, right?
<claytonc> yes
<claytonc> we do end to end TLS and have since day one
<semushin> claytonc, I'm not sure that the doc readers will understand the differences between rest and REST. I think that "encrypted at the datastore layer" is more obvious and don't allow to be misinterpreted. Can I use it in our docs instead of "at rest"?
<claytonc> yes
```

PTAL again.

File: admin_guide/encrypting_data.adoc [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+|===
+
+Each provider supports multiple keys. The keys are tried in order for decryption, and if the
+provider is the first provider, the first key is used for encryption.
+
+[NOTE]
+====
+Kubernetes has no proper nonce generator and uses a random IV as nonce for AES-GCM. Since AES-GCM
+requires a proper nonce to be secure, AES-GCM is not recommended. The 200k write limit just limits
+the possibility of a fatal nonce misuse to a reasonable low margin.
```
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: @tiran @simo5 I've <strike>added</strike> copied @tiran quote about why we don't recommend AES-GCM. Please, check that it's ok.

Comment (simo5) [inline admin_guide/encrypting_data.adoc]: Looks good, thanks!

**Comment by php-coder** [comment]
@openshift/team-documentation What else I need to do to make it merged?  If it's possible I'd like to have it merged ASAP as it's blocking my Trello card that is required to be completed for the upcoming release. I can fix possible issues in this doc as follow-ups. Thanks!

File: admin_guide/encrypting_data.adoc [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+:toc: macro
+:toc-title:
+
+toc::[]
+
+[[encrypting-data-overview]]
+== Overview
+
+This topic shows how to enable and configure encryption of secret data at datastore layer. While
+examples use the `secrets` resource, any resource can be encrypted. Yet another candidate is
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: any resource can be encrypted, such as `*ConfigMaps*`.

File: admin_guide/encrypting_data.adoc [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+        - name: key2
+          secret: dGhpcyBpcyBwYXNzd29yZA==
+    - identity: {}
+----
+<1> Each `resources` array item is a separate config and contains a complete configuration.
+<2> The `resources.resources` field is an array of Kubernetes resource names (*resource* or
+*resource.group*) that should be encrypted.
+<3> The `providers` array is an ordered xref:encrypting-data-providers[list of
+the possible encryption providers]. Only one provider type may be specified per entry (*identity* or
+*aescbc* may be provided, but not both in the same item).
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: Will our users know what aescbc is?

Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: IMHO if they decided to encrypt the data they probably already know what it is or they don't even need to know the details and just follow our recommendations.

File: admin_guide/encrypting_data.adoc (hunk 18 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: Fixed; but I used `configmaps` instead of `ConfigMaps` because I'm not sure whether it works in this way.

**Comment by php-coder** [comment]
@mburke5678 Thanks! Comments have been addressed.

File: admin_guide/encrypting_data.adoc [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+| *Is not recommended* for use except when an automated key rotation scheme is implemented.
+
+|===
+
+Each provider supports multiple keys. The keys are tried in order for decryption, and if the
+provider is the first provider, the first key is used for encryption.
+
+[NOTE]
+====
+Kubernetes has no proper nonce generator and uses a random IV as nonce for AES-GCM. Since AES-GCM
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: Will our users know what `nonce` and `IV` are? I cannot find other references to these terms in the docs. 
Should probably spell out  `initialization vector (IV)` on first reference

File: admin_guide/encrypting_data.adoc [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+----
+
+. Place that value in the *secret* field
+
+. Restart API server
+
+[IMPORTANT]
+====
+Config file contains keys that can decrypt content in etcd, so you must properly restrict
+permissions on masters so only the user who runs the master API server can read it.
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: s/Config file/A config file (or) The config file. 
Do we need to give the name and location of the config file?

File: admin_guide/encrypting_data.adoc [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+====
+
+
+[[encrypting-data-verification]]
+== Verifying that Data is Encrypted
+
+Data is encrypted when written to etcd. After restarting API server any newly created or
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: s/server/the server

File: admin_guide/encrypting_data.adoc [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+----
+ETCDCTL_API=3 etcdctl get /kubernetes.io/secrets/default/secret1 -w fields \
+--cacert=/var/lib/origin/openshift.local.config/master/ca.crt \
+--key=/var/lib/origin/openshift.local.config/master/master.etcd-client.key \
+--cert=/var/lib/origin/openshift.local.config/master/master.etcd-client.crt \
+--endpoints 'https://127.0.0.1:4001' | grep Value
+----
+
+. Verify the stored secret is prefixed with *k8s:enc:aescbc:v1:* which indicates the *aescbc*
+provider has encrypted the resulting data.
```
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: How do we verify?

**Comment by mburke5678** [comment]
@php-coder Few comments. I can do the formatting/grammar in a follow-up PR after merge, if you want.

File: admin_guide/encrypting_data.adoc (hunk 23 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: It supposed that the previous command returned a secret as raw data and here we're verifying that this raw representation "is prefixed with *k8s:enc:aescbc:v1:*".

File: admin_guide/encrypting_data.adoc (hunk 22 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: @mburke5678 did you mean s/API server/the API server/?

File: admin_guide/encrypting_data.adoc (hunk 21 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: > Do we need to give the name and location of the config file?

We don't know the name and location because it's up to the admin where he will put it. I'm going to rephrase it to "The encryption provider config file contains ..."

File: admin_guide/encrypting_data.adoc (hunk 20 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: It is supposed that this note explains some deep knowledge that targets and interesting to the people who deeply understand what these terms mean.

File: admin_guide/encrypting_data.adoc (hunk 22 above)
Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: @mburke5678 All comments except this one were addressed.

File: admin_guide/encrypting_data.adoc (hunk 23 above)
Comment (mburke5678) [inline admin_guide/encrypting_data.adoc]: Sorry, what is the process or command to verify? Do we need that in the docs?

Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: The command where in the previous point:
```
. Using the `etcdctl` commandline, read that secret out of etcd:
+
----
ETCDCTL_API=3 etcdctl get /kubernetes.io/secrets/default/secret1 -w fields [...] | grep Value
----
```
But the paragraph was a long so it's possible that reader may lose the context.

I'll replace "Verify the stored secret is ..." by "Verify that the output of the command above is ..."

Comment (php-coder) [inline admin_guide/encrypting_data.adoc]: Done. PTAL.

**Comment by php-coder** [comment]
@openshift/team-documentation Ping.

File: admin_guide/encrypting_data.adoc (hunk 22 above)
Comment (mburke5678) [reply admin_guide/encrypting_data.adoc]: @php-coder Yes. the API server.

**Comment by mburke5678** [comment]
@php-coder Which OpenShift versions does this PR apply to? 3.4, 3.5, 3.6?? Is this set for immediate release?

File: admin_guide/encrypting_data.adoc (hunk 22 above)
Comment (php-coder) [reply admin_guide/encrypting_data.adoc]: Ok, fixed.

**Comment by php-coder** [comment]
> Which OpenShift versions does this PR apply to? 3.4, 3.5, 3.6?? Is this set for immediate release?

@mburke5678 It's a new feature that is available in 3.6 only

File: _topic_map.yml (hunk 1 above)
Comment (php-coder) [reply _topic_map.yml]: @mburke5678 Do you know something about ^^^ ?

Comment (mburke5678) [reply _topic_map.yml]: Per @adellape Whether a doc goes in Dedicated "it usually depends on whether the "dedicated-cluster-admin" role can operate on whatever the topic is talking about. see: https://docs.openshift.com/dedicated/admin_guide/index.html"
Also, "if it does require modifying the master_config or node_config files, probably not relevant to OSD."
Your file required editing master config, so it would not go in Dedicated.

Comment (php-coder) [reply _topic_map.yml]: Ok, thanks! I've pushed an update.

**Comment by ahardin-rh** [comment]
Thank you! I will follow up with some very minor style edits in a separate PR.

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../ admin_guide/encrypting_data.adoc#admin-guide-encrypting-data-at-datastore[Encrypting Data at Datastore Layer]
|New topic on how to enable and configure encryption of secret data at the datastore layer.
%
